### PR TITLE
Quick check for git commit hook

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

Previously, we always notice checkstyle and CodeNarc errors only when pushed to CI. This has a negative impact on developer productivity. This PR adds a `quickCheck` tasks which runs checkstyle/CodeNarc/ktlin check only on changed files since last commit.

For a warmed-up daemon, it takes 4~5s to check a changed file:

```
$ time ./gradlew quickCheck 
> Task :quickCheck FAILED
CodeNarc Report - 10 Jan 2020, 11:18:45

Summary: TotalFiles=6 FilesWithViolations=1 P1=0 P2=0 P3=1

File: CheckstylePluginIntegrationTest.groovy
    Violation: Rule=UnusedImport P=3 Line=18 Msg=[The [java.util.concurrent.ConcurrentMap] import is never referenced] Src=[import java.util.concurrent.ConcurrentMap]

[CodeNarc (http://www.codenarc.org) v1.5]
CodeNarc completed: (p1=0; p2=0; p3=1) 891ms
ERROR: Number of p3 violations greater than maximum of 0

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quickCheck'.
> Process 'command '/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java'' finished with non-zero exit value 1

BUILD FAILED in 4s
1 actionable task: 1 executed
./gradlew --dependency-verification off quickCheck  1.57s user 0.22s system 33% cpu 5.318 total


$ time ./gradlew quickCheck 
> Task :quickCheck FAILED
Starting audit...
[ERROR] /Users/zhb/Projects/gradle/subprojects/core-api/src/main/java/org/gradle/api/Project.java:60:8: Unused import - java.util.concurrent.ConcurrentMap. [UnusedImports]
Audit done.
Checkstyle ends with 1 errors.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quickCheck'.
> Process 'command '/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java'' finished with non-zero exit value 1

BUILD FAILED in 3s
1 actionable task: 1 executed
./gradlew --dependency-verification off quickCheck  1.48s user 0.20s system 39% cpu 4.296 total
```

This commit adds a quickCheck task for gradle which only checks changed
files since last commit. It usually takes 5s for a warmed-up daemon.

### CODE REVIEW

File: gradle/verification-metadata.xml [hunk 1]
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<verification-metadata xmlns="https://schema.gradle.org/dependency-verification"
-                       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-                       xmlns:schemaLocation="https://schema.gradle.org/dependency-verification https://schema.gradle.org/dependency-verification/verification-1.0.xsd">
+<verification-metadata xmlns="https://schema.gradle.org/dependency-verification" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:schemaLocation="https://schema.gradle.org/dependency-verification https://schema.gradle.org/dependency-verification/dependency-verification-1.0.xsd">
```
Comment (blindpirate) [inline gradle/verification-metadata.xml]: @melix this is formatted automatically by `./gradlew -M`, so I think I'd better check it in.

File: buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        checks.forEach { it.addDependencies(project) }
+        checks.forEach { it.runCheck(project, it.filter(changedFiles)) }
+    }
+
+    //    A       buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt
+    //    M       gradle.properties
+    //    D       gradle/Check.java
+    private
+    fun getChangedFiles(): List<String> =
+        project.execAndGetStdout("git", "diff", "--cached", "--name-status", "HEAD")
```
Comment (melix) [inline buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: Would using `jgit` make sense here instead of relying on CLI?

File: gradle/verification-metadata.xml (hunk 1 above)
Comment (melix) [reply gradle/verification-metadata.xml]: the automatic formatter cannot generate the standard multi-line thing so I would just ignore those changes and keep the multiline thing.

File: gradle/verification-metadata.xml [hunk 3]
```diff
@@ -1111,6 +1109,11 @@
             <pgp value="9e93db1192cee3d3b581aabb37e9d58ea4598641"/>
          </artifact>
       </component>
+      <component group="org.javassist" name="javassist" version="3.24.0-GA">
+         <artifact name="javassist-3.24.0-GA.jar">
+            <pgp value="666a4692ce11b7b3f4eb7b3410066a9707090cf9"/>
+         </artifact>
+      </component>
```
Comment (jjohannes) [inline gradle/verification-metadata.xml]: Why is this needed?

**Comment by blindpirate** [comment]
@melix No, it's more convenient. For exec tasks I think the most expensive operation is forking a process, so using a real task or not doesn't matter, right?

File: buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt (hunk 2 above)
Comment (blindpirate) [reply buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: `jgit` would be several hundred ms faster. But CLI is more readable and maintable. So let's keep CLI unless it becomes a problem.

File: gradle/verification-metadata.xml (hunk 3 above)
Comment (blindpirate) [reply gradle/verification-metadata.xml]: That would be a question for @melix .... Should I check in this change?

**Comment by blindpirate** [comment]
@jjohannes I can sure make follow up issue to make the "CodeNarc" and "Checkstyle" tasks incremental, but in this case, I think git changed files might be more suitable since that's how standard git commit hook works.

I was trying incremental tasks at first, but that means UP-TO-DATE check for all java/groovy files to find out which files are changed, right? I'm not sure that would be faster. @wolfs What do you think here?

File: gradle/verification-metadata.xml (hunk 3 above)
Comment (melix) [reply gradle/verification-metadata.xml]: You should tell me ;) If it works _without_ you shouldn't. If it works with it, you should check the signature and add it.

File: buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt (hunk 2 above)
Comment (wolfs) [reply buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: I suppose you should use `--porcelain`, so the output can be parsed reliably.

**Comment by wolfs** [comment]
@blindpirate VFS retention (aka file watching to detect changes) already works and we are going to make things faster there as well. So the up-to-date checks for the sources don't need to check all the files, only the directories that were changed.   So I think this can become very fast and we wouldn't need to rely on git for change detection.

**Comment by vlsi** [comment]
> Can you create follow up issue to make the "CodeNarc" and "Checkstyle" tasks incremental? 

Just in case, my experience shows that adding `cacheFile` property to Checkstyle configuration significantly improves task performance: https://github.com/apache/jmeter/commit/447e4ce3d44a4fce2d4948ed8777ff2fae5ebddc

Have you tried that?

File: buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt (hunk 2 above)
Comment (blindpirate) [reply buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: Maybe I misunderstand you, but `git status --porcelain` has same output as `git diff --cached --name-status HEAD`:

```
M build.gradle.kts
A some-new-file
```

Why `--porcelain` are more reliable?

**Comment by blindpirate** [comment]
@vlsi Thanks, will try.

File: buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt (hunk 2 above)
Comment (wolfs) [reply buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: This is the documentation: https://git-scm.com/docs/git-status#Documentation/git-status.txt---p

The "porcelain" commands (like `checkout`, `pull`, `commit`, etc.) have no guarantee for stable output between versions. The `--porcelain` command line flag adds this guarantee, so the output is easy to parse for scripts.

Comment (blindpirate) [reply buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt]: Thanks @wolfs
